ops/charm.py

```diff
@@ -1448,6 +1449,29 @@ def _evaluate_status(charm: CharmBase):  # pyright: ignore[reportUnusedFunction]
         unit.status = model.StatusBase._get_highest_priority(unit._collected_statuses)
 
 
+def _send_databag_to_juju(charm: CharmBase):  # pyright: ignore[reportUnusedFunction]
+    for (relation_id, entity), (obj, _, encoder) in charm.model._cache._databag_obj_cache.items():
+        if encoder is None:
+            encoder = json.dumps
+        if hasattr(obj, '__dataclass_fields__'):
+            # A standard library dataclass or Pydantic dataclass.
+            fields = (field.name for field in dataclasses.fields(obj))
+            values = dataclasses.asdict(obj)
+        elif hasattr(obj, 'model_fields'):
+            # A Pydantic model.
+            fields = obj.model_fields
+            values = obj.model_dump()
```

Comment:
For what it's worth, the obs charms pass in extra args here, `mode="json", by_alias=True, exclude_defaults=True`